### PR TITLE
IDEM-246-Jazz

### PR DESCRIPTION
**Pull request by JasminFarhi**

### CODE REVIEW

File: src/data/migrations/CNew-03-request.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+BEGIN;
+CREATE TABLE IF NOT EXISTS "testers" (
+    "testerId" SERIAL,
+    "email" VARCHAR NOT NULL,
```
Comment (bitcoinbrisbane) [inline src/data/migrations/CNew-03-request.sql:4]: Can we make email the primary key?

**Review by bitcoinbrisbane — COMMENTED** [review]
One thing to consider

File: src/data/migrations/CNew-03-request.sql (hunk 1 above)
Comment (xiangzhe0812) [reply src/data/migrations/CNew-03-request.sql:4]: Id as PK and make email unique.

Comment (JasminFarhi) [reply src/data/migrations/CNew-03-request.sql:4]: The email is unique, and the user ID is the PK. :)

File: src/data/entities/tester.entity.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    testerId: number;
+
+    @Column({ unique: true })
+    email: string;
+
+    @Column({ unique: true })
+    firstName: string;
+
+    @Column({ unique: true })
+    lastName: string;
```
Comment (xiangzhe0812) [inline src/data/entities/tester.entity.ts:20]: first name and last name should not be unique

File: src/main.ts [hunk 3]
```diff
@@ -25,6 +25,6 @@ async function bootstrap() {
 
     SwaggerModule.setup("swagger", app, document);
 
-    await app.listen(3000);
+    await app.listen(3001);
```
Comment (xiangzhe0812) [inline src/main.ts:28]: can we roll back to 3000?

File: src/user/user.controller.ts [hunk 4]
```diff
@@ -169,4 +171,16 @@ export class UserController {
...
+
+    @Post("tester")
+    @ApiOperation({ summary: "request testflight tester" })
+    @ApiResponse({
+        status: HttpStatus.OK
+    })
+    @ApiResponse({
+        status: HttpStatus.BAD_REQUEST
+    })
+    async tester(@Body() body: TestFlightRequest): Promise<Tester> {
```
Comment (xiangzhe0812) [inline src/user/user.controller.ts:183]: can we make the function name better?
such as requestTest or applyTest

File: src/user/user.controller.ts [hunk 5]
```diff
@@ -169,4 +171,16 @@ export class UserController {
...
+        return this.userService.testflightTester(body);
```
Comment (xiangzhe0812) [inline src/user/user.controller.ts:184]: same here, maybe this.userService.requestBeTester

File: src/user/user.service.ts [hunk 6]
```diff
@@ -60,6 +63,19 @@ export class UserService {
...
     }
 
+    public async testflightTester(
+        testFlightRequest: TestFlightRequest
+    ): Promise<Tester> {
+        const tester = await this.testerRepository.findOneBy({
+            email: testFlightRequest.email.toLowerCase()
+        });
+        if (!tester) {
+            this.logger.verbose(`New user ${testFlightRequest.email} created`);
```
Comment (xiangzhe0812) [inline src/user/user.service.ts:73]:
```suggestion
            this.logger.verbose(`New tester ${testFlightRequest.email} created`);
```

**Review by bitcoinbrisbane — DISMISSED** [review]

File: src/interfaces.ts [hunk 7]
```diff
@@ -100,8 +101,10 @@ export interface IKycService {
 }
 
 export interface IUserService {
+    requestToBeTester(body: TestFlightRequest): Tester | PromiseLike<Tester>;
     verify(body: UserVerifyRequestBody): Promise<string>;
     findOne(email: string): Promise<User>;
+    testflightTester(body: TestFlightRequest): Promise<Tester>;
```
Comment (xiangzhe0812) [inline src/interfaces.ts:107]: this guy no longer used

File: src/interfaces.ts [hunk 8]
```diff
@@ -100,8 +101,10 @@ export interface IKycService {
 }
 
 export interface IUserService {
+    requestToBeTester(body: TestFlightRequest): Tester | PromiseLike<Tester>;
```
Comment (xiangzhe0812) [inline src/interfaces.ts:104]:
```suggestion
    requestToBeTester(body: TestFlightRequest):  Promise<Tester>;
```

File: src/data/entities/tester.entity.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+@Entity("testers")
+export class Tester {
+    @PrimaryGeneratedColumn()
+    testerId: number;
+
+    @Column({ unique: true })
+    email: string;
+
+    @Column({ unique: false })
```
Comment (pondpiu) [inline src/data/entities/tester.entity.ts:16]: unqiue is false by default
i think its clearer to only put unique: true 
and leave others as default false value

**Review by xiangzhe0812 — APPROVED** [review]

File: src/interfaces.ts [hunk 10]
```diff
@@ -100,6 +101,7 @@ export interface IKycService {
 }
 
 export interface IUserService {
+    requestToBeTester(body: TestFlightRequest): PromiseLike<Tester>;
```
Comment (pondpiu) [inline src/interfaces.ts:104]: Why PromiseLike

File: src/user/user.service.ts [hunk 11]
```diff
@@ -60,6 +63,21 @@ export class UserService {
         return user;
     }
 
+    public async requestToBeTester(
+        testFlightRequest: TestFlightRequest
+    ): Promise<Tester> {
+        const requestTest = await this.testerRepository.findOneBy({
+            email: testFlightRequest.email.toLowerCase()
+        });
+        if (!requestTest) {
```
Comment (pondpiu) [inline src/user/user.service.ts:72]: can be made cleaner using guard clause & early return

if(requestTest) return requestTest

... rest

File: src/user/user.service.ts [hunk 12]
```diff
@@ -60,6 +63,21 @@ export class UserService {
...
+            this.logger.verbose(
+                `New tester ${testFlightRequest.email} created`
+            );
+            return this.testerRepository.save(testFlightRequest);
```
Comment (pondpiu) [inline src/user/user.service.ts:76]: should log after user is created, otherwise log will show up even .save() fail

`await repo.save()
log()
`

**Review by bitcoinbrisbane — APPROVED** [review]

**Review by pondpiu — APPROVED** [review]
LGTM

**Pull request merged**